### PR TITLE
add upgrade case to test clusterlogforwarder

### PR DESCRIPTION
**Pull request by QiaolingTang**
case: https://url.corp.redhat.com/f0086ba
upgrade-prepare: https://url.corp.redhat.com/b72f05c
upgrade-check: https://url.corp.redhat.com/5b5458d

/cc @anpingli @gkarager

### CODE REVIEW

**Review by anpingli — COMMENTED** [review]
/lgtm

**Comment by QiaolingTang** [comment]
New logs:
upgrade-prepare: https://url.corp.redhat.com/3703f34
upgrade-check: https://url.corp.redhat.com/4765d59

@anpingli , sorry, I updated this PR, could you please review it again? Thanks.

File: features/step_definitions/logging.rb [hunk 1]
```diff
@@ -648,48 +661,67 @@
...
-  success = wait_for(600, interval: 10) {
-    elasticsearch('elasticsearch').cluster_health == "green" &&
-    (elasticsearch('elasticsearch').nodes_status.last["upgradeStatus"].empty? ||
-    elasticsearch('elasticsearch').nodes_status.last["upgradeStatus"]["scheduledUpgrade"].nil?)
-  }
-  raise "ES cluster isn't in a good status" unless success
-  # check pvc count
-  unless BushSlicer::PersistentVolumeClaim.list(user: user, project: project).count == cluster_logging('instance').logstore_node_count
-    raise "The PVC count doesn't match the ES node count"
+  if cluster_logging('instance').log_store_spec != nil
```
[the remark this answers is not in the document]
Comment (QiaolingTang) [reply features/step_definitions/logging.rb:664]: If the patch command fails, then the scenario will be failed. https://github.com/openshift/verification-tests/pull/1935/files#diff-25b7b9fb37c470ca782e7381a157ec7648e98a1a80f61f2cce948bba99bf676cR649 .

The upgrade only happens when the channel or/and catalog source changed, if there  is no change, then the step `I upgrade the operator with` won't be executed: https://github.com/openshift/verification-tests/pull/1935/files#diff-25b7b9fb37c470ca782e7381a157ec7648e98a1a80f61f2cce948bba99bf676cR724.

I didn't process the condition when the catalog source changed, I'll modify it.

[the remark this answers is not in the document]
Comment (QiaolingTang) [reply features/step_definitions/logging.rb:664]: The steps in upgrade-check scenario are:
1. check if logging could work as expected after cluster is upgraded
2. check if logging should to be upgraded, if yes, upgrade logging and ensure EFK pods can be upgraded
3. check if logging could work as expected
There are some duplication in step 1 and step 3 if step 2 is not executed, but I think this condition won't happen frequently.

**Comment by QiaolingTang** [comment]
Logs:
OCP-29743:
upgrade-prepare: https://url.corp.redhat.com/d042d73
upgrade-check: https://url.corp.redhat.com/8cea59a

/hold cancel

**Comment by QiaolingTang** [comment]
@anpingli could you please help to review this PR again?

**Review by anpingli — COMMENTED** [review]
/lgtm

File: lib/openshift/cluster_logging.rb [hunk 2]
```diff
@@ -124,19 +124,19 @@ def wait_until_kibana_is_ready(user: nil, quiet: false, timeout: 10*60)
       end
     end
 
-    private def collection_spec(user: nil, quiet: false, cached: false)
+    def collection_spec(user: nil, quiet: false, cached: false)
```
Comment (pruan-rht) [inline lib/openshift/cluster_logging.rb:127]: @QiaolingTang any reason you changed these methods from `private`?

Comment (QiaolingTang) [reply lib/openshift/cluster_logging.rb:127]: @pruan-rht In clusterlogging, the collection_spec, log_store_spec, curation_spec and visualization_spec are optional. In some scenarios we may not define all of them when creating clusterlogging/instance. I want to use these methods to check if these specs are defined or not. An example: https://github.com/openshift/verification-tests/pull/1935/files#diff-25b7b9fb37c470ca782e7381a157ec7648e98a1a80f61f2cce948bba99bf676cR667

Comment (pruan-rht) [reply lib/openshift/cluster_logging.rb:127]: /lgtm

**Pull request merged**